gateway/blocks_gateway.go

```diff
@@ -223,38 +231,215 @@ func (api *BlocksGateway) Head(ctx context.Context, path ImmutablePath) (Content
 		errCh <- multierr.Combine(carWriteErr, pipeCloseErr)
 		close(errCh)
 	}()
 
-	return md, r, errCh, nil
+	return r, errCh, nil
+}
+
+// walkGatewaySimpleSelector walks the subgraph described by the path and terminal element parameters
+func walkGatewaySimpleSelector(ctx context.Context, p ipfspath.Path, params CarParams, lsys *ipld.LinkSystem, pathResolver resolver.Resolver) error {
+	// First resolve the path since we always need to
+	lastCid, remainder, err := pathResolver.ResolveToLastNode(ctx, p)
+	if err != nil {
+		return err
+	}
+
+	// If we just need to resolve the path then we're done
+	if params.Scope == dagScopeBlock {
+		return nil
+	}
+
+	lctx := ipld.LinkContext{Ctx: ctx}
+	pathTerminalCidLink := cidlink.Link{Cid: lastCid}
+
+	// If we're asking for everything then give it
+	if params.Scope == dagScopeAll {
+		lastCidNode, err := lsys.Load(lctx, pathTerminalCidLink, basicnode.Prototype.Any)
+		if err != nil {
+			return err
+		}
+
+		sel, err := selector.ParseSelector(selectorparse.CommonSelector_ExploreAllRecursively)
+		if err != nil {
+			return err
+		}
+
+		progress := traversal.Progress{
+			Cfg: &traversal.Config{
+				Ctx:        ctx,
+				LinkSystem: *lsys,
+			},
+		}
+
+		if err := progress.WalkMatching(lastCidNode, sel, func(progress traversal.Progress, node datamodel.Node) error {
+			return nil
+		}); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Since we need more of the graph load it to figure out what we have
+	// This includes determining if the terminal node is UnixFS or not
+	pc := dagpb.AddSupportToChooser(func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+
+	np, err := pc(pathTerminalCidLink, lctx)
+	if err != nil {
+		return err
+	}
+
+	lastCidNode, err := lsys.Load(lctx, pathTerminalCidLink, np)
+	if err != nil {
+		return err
+	}
+
+	if pbn, ok := lastCidNode.(dagpb.PBNode); !ok {
+		// If it's not valid dag-pb then we're done
+		return nil
+	} else if len(remainder) > 0 {
+		// If we're trying to path into dag-pb node that's invalid and we're done
+		return nil
+	} else if !pbn.FieldData().Exists() {
+		// If it's not valid UnixFS then we're done
+		return nil
+	} else if unixfsFieldData, decodeErr := data.DecodeUnixFSData(pbn.Data.Must().Bytes()); decodeErr != nil {
+		// If it's not valid dag-pb and UnixFS then we're done
+		return nil
+	} else {
+		switch unixfsFieldData.FieldDataType().Int() {
+		case data.Data_Directory, data.Data_Symlink:
+			// These types are non-recursive so we're done
+			return nil
+		case data.Data_Raw, data.Data_Metadata:
+			// TODO: Is Data_Raw actually different from Data_File or is that a fiction?
+			// TODO: Deal with UnixFS Metadata being handled differently in boxo/ipld/unixfs than go-unixfsnode
+			// Not sure what to do here so just returning
+			return nil
```

Comment:
Some TODOs here. IIUC both of these are dealt with differently in go-unixfsnode vs boxo/ipld/unixfs. Need to figure out what to do here... and have some fixtures to illustrate.

@Jorropo @rvagg @ribasushi any insight you can give me that might save me time poking around here?

Reply:
I have yet to see a usecase where `Raw` is different from `File`.
In my spec PR I say it's a file with a metadata like links that is hard coded to nothing and length that is derived from the block lengths.

Reply:
unixfsnode handles raw like a file and metadata like a symlink: https://github.com/ipfs/go-unixfsnode/blob/f4d6da0dbdf3614269b3580eeebe2bb5245f09bc/reification.go#L54-L87

I think those are up for changing if the spec clarifies it differently since I don't believe they're particularly common and the kinds of changes we're likely to make make probably won't show up under common use of those code paths.

Reply:
like a symlink - essentially a unixfs pathable structure, i.e. directory

Reply:
> I think for now we can return a 501 Not Implemented.

IIUC we can't do this in a meaningful way, we're sending out a CAR file stream so are back in lack of HTTP streaming errors/trailers land. 

So this seems to imply just returning (like we do with basic directories).

There is some difficulty here around upgrading in the future to support traversals where we didn't before and figuring out which gateways have upgraded vs not. Might just be a future problem though 🤷.

Reply:
> I have yet to see a usecase where Raw is different from File.

So this means we should recurse into both of them, right? @rvagg seems to also imply switching the `TRaw` preloader to be recursive in go-unixfsnode. Sound good?

Don't feel so strongly about what we do as long as we're all doing the same thing 😄

Reply:
I'm closing this for now, left a comment in-code. This can be handled later.